dipy/tracking/utils.py

```diff
@@ -568,6 +573,122 @@ def target(streamlines, target_mask, affine, include=True):
             yield sl
 
 
+def streamline_near_roi(sl, x_roi_coords, tol, mode='any'):
+    """
+    Is a streamline near an ROI.
+
+    Implements the inner loops of the :func:`near_roi` function.
+
+    Parameters
+    ----------
+    sl : array, shape (N, 3)
+	    A single streamline
+    x_roi_coords : array, shape (M, 3)
+        ROI coordinates transformed to the streamline coordinate frame.
+    tol : float
+        Distance (in the units of the streamlines, usually mm). If any
+        coordinate in the streamline is within this distance from the center
+        of any voxel in the ROI, this function returns True.
+	mode : string
+		One of {"any", "all", "either_end", "both_end"}, where return True if:
+		"any" : any point is within tol from ROI.
+		"all" : all points are within tol from ROI.
+		"either_end" : either of the end-points is within tol from ROI
+		"both_end" : both end points are within tol from ROI.
+
+	Returns
+	-------
+	out : boolean
+    """
+    if len(x_roi_coords)==0:
+        return False
+    if mode=="any" or mode=="all":
+        s = sl
+    elif mode=="either_end" or mode=="both_end":
+        # 'end' modes, use a streamline with 2 nodes:
+        s = np.vstack([sl[0], sl[-1]])
+    else:
+        e_s = "For determining relationship to an array, you can use "
+        e_s += "one of the following modes: 'any', 'all', 'both_end',"
+        e_s += "'either_end', but you entered: %s."%mode
+        raise ValueError(e_s)
+
+    dist = cdist(s, x_roi_coords, 'euclidean')
+
+    if mode=="any" or mode=="either_end":
+        return np.min(dist)<=tol
+    else:
+        return np.all(np.min(dist, -1)<=tol)
+
+
+def near_roi(streamlines, target_mask, affine=None, tol=None,
+             mode="any"):
+    """
+    Provide filtering criteria for a set of streamlines based on whether they
+    fall within a tolerance distance from an ROI
+
+    Parameters
+    ----------
+    streamlines : list or generator
+        A sequence of streamlines. Each streamline should be a (N, 3) array,
+        where N is the length of the streamline.
+    target_mask : ndarray
+        A mask used as a target. Non-zero values are considered to be within
+        the target region.
+    affine : ndarray
+        Affine transformation from voxels to streamlines. Default: identity.
+    tol : float
+        Distance (in the units of the streamlines, usually mm). If any
+        coordinate in the streamline is within this distance from the center
+        of any voxel in the ROI, the filtering criterion is set to True for
+        this streamline, otherwise False. Defaults to the distance between
+        the center of each voxel and the corner of the voxel.
+	mode : string, optional
+		One of {"any", "all", "either_end", "both_end"}, where return True if:
+		"any" : any point is within tol from ROI. Default.
```

Comment:
This will probably not render correctly. Because ``:`` is reserved in the docstrings for parameters.